Refactor benchmarks

fdcastel opened the pull request:
_Based on #1200 for now. I will rebase it with latest `master` in the future._

- Rename project `Perf` to `FirebirdSql.Data.FirebirdClient.Benchmarks`.
- Update project to use .NET8.
- Upgrade BenchmarkDotNet to version `0.14.0`.
- Update baseline nuget package to `v10.3.1`.
- Add `/BenchmarkDotNet.Artifacts` folder to `.gitignore`.
- Pass command-line arguments to BenchmarkDotNet engine.
- Apply SQL Formatting. Use raw strings.
- Add script `run-benchmark.ps1`.

fdcastel committed:
Adds FbTransactionInfo.GetTransactionId().

fdcastel committed:
Adds .NET distributed tracing instrumentation.

fdcastel committed:
Add metrics.

fdcastel committed:
Refactor: Benchmarks.

- Rename project 'Perf' to 'FirebirdSql.Data.FirebirdClient.Benchmarks'.
- Update project to use .net8.
- Upgrade BenchmarkDotNet to version 0.14.0.
- Update baseline nuget package to v10.3.1.
- Add /BenchmarkDotNet.Artifacts to .gitignore.
- Pass command-line arguments to BenchmarkDotNet engine.
- Apply SQL Formatting. Use raw strings.
- Add script run-benchmark.ps1.

willibrandon commented on src/FirebirdSql.Data.FirebirdClient.Benchmarks/CommandBenchmark.cs:
```diff
@@ -34,30 +34,46 @@ public Config()
-			AddJob(baseJob.WithCustomBuildConfiguration("ReleaseNuGet").WithId("NuGet").AsBaseline());
+
+			AddValidator(BaselineValidator.FailOnError);
+			AddValidator(JitOptimizationsValidator.FailOnError);
```
If full JSON (*full.json) reports are exported then benchmarks results can be compared historically using [ResultsComparer](https://github.com/dotnet/performance/tree/main/src/tools/ResultsComparer). 

`AddExporter(JsonExporter.Full);`

willibrandon replied:
Never mind, I see how it is supposed to work now!

fdcastel committed:
Updates run-benchmark.ps1.

fdcastel commented:
Rebase with `master`.

fdcastel committed:
Fix NullReferenceException when StartActivity returns null

StartActivity() returns null when no listeners are interested in the
activity. The code accessed activity.IsAllDataRequested without a null
check, causing a NullReferenceException. Use the Npgsql-style pattern
'activity is not { IsAllDataRequested: true }' for early return.

fdcastel committed:
Fix Stopwatch elapsed time conversion

Stopwatch ticks are not TimeSpan ticks when Stopwatch.IsHighResolution
is true (most systems). TimeSpan.FromTicks(stopwatchTicks) produces
incorrect durations. Use Stopwatch.GetElapsedTime() which correctly
handles the frequency conversion.

fdcastel committed:
Rename db.system to db.system.name per OTel semantic conventions v1.40+

The semantic conventions renamed db.system to db.system.name. The
well-known value for Firebird is 'firebirdsql', not 'firebird'.
Updated in both FbActivitySource (spans) and FbConnection (metrics).

fdcastel committed:
Set db.namespace attribute on spans from connection database name

The semantic conventions mark db.namespace as Conditionally Required
when available. The connection's Database property is used as the value.

fdcastel committed:
Set error.type attribute on failure spans

The semantic conventions mark error.type as Conditionally Required on
failure. For FbException, use the SQLSTATE code; for other exceptions,
use the full exception type name.

fdcastel committed:
Gate db.query.text and db.query.parameter.* behind opt-in flags

Per OTel semantic conventions, db.query.text and db.query.parameter.*
are Opt-In level attributes that may expose sensitive data. They should
not be collected by default. Added IsQueryTextTracingEnabled flag and
reuse IsParameterLoggingEnabled for parameter tracing.

fdcastel committed:
Read ActivitySource/Meter version from assembly metadata

Replace hardcoded '1.0.0' version with the assembly's informational
version, keeping the telemetry version in sync with the NuGet package.

fdcastel committed:
Fix activity lifecycle: record metrics on error path, prevent stale recording

TraceCommandException now records metrics immediately and resets
_startedAtTicks so the later TraceCommandStop (called during Release)
does not record a second, inflated duration that includes cleanup time.

fdcastel committed:
Remove non-standard and deprecated telemetry attributes

- Remove db.transaction_id custom tag (not in OTel spec, causes server
  round-trip per traced command)
- Remove commented-out db.snapshot_id code
- Remove deprecated exception.escaped attribute
- Remove NormalizeDbNull helper, inline DBNull check

fdcastel committed:
Add server.port, db.stored_procedure.name; fix server.address in metrics

- Set server.port on spans when using a non-default port
- Set db.stored_procedure.name for StoredProcedure command types
- Fix server.address in MetricsConnectionAttributes to not concatenate
  the port (server.port is now a separate attribute)
- Clean up leftover placeholder comments

fdcastel committed:
Add db.query.summary and db.operation.name for Text commands

Set db.query.summary as the low-cardinality activity name on all spans.
For Text commands, extract the first SQL verb (SELECT, INSERT, etc.)
and use it as db.operation.name and in the activity name.

fdcastel committed:
Add debug assertions in TraceCommandStop and TraceCommandException

Debug.Assert checks that TraceCommandStart was called before the stop
or exception path, helping catch lifecycle bugs during development.

fdcastel committed:
Initialize MetricsConnectionAttributes to empty array

Prevents potential NullReferenceException if metrics are recorded
before the connection string is set.

fdcastel committed:
Avoid dictionary allocation in observable metric callbacks

Change GetMetrics() to return IEnumerable of tuples via Select instead
of allocating a Dictionary via ToDictionary on every collection cycle.
Also simplify GetConnectionMax to use Select instead of SelectMany.

fdcastel committed:
Add InstrumentAdvice with histogram bucket boundaries on .NET 9+

Suggest default histogram bucket boundaries for operation duration and
connection create time histograms, improving out-of-the-box resolution.
Guarded by NET9_0_OR_GREATER since InstrumentAdvice is a .NET 9 API.

fdcastel committed:
Add FbTelemetry public class with ActivitySource and Meter names

Expose the telemetry source names as public constants so consumers can
easily subscribe: builder.AddSource(FbTelemetry.ActivitySourceName) and
builder.AddMeter(FbTelemetry.MeterName). Internal code updated to use
the constants.

fdcastel committed:
Add OpenTelemetry integration documentation

Document the ActivitySource and Meter names, available span attributes,
opt-in sensitive attributes, and available metrics with their types and
semantic convention references.

fdcastel committed:
Fix IndexOfAny compilation on all target frameworks

fdcastel committed:
Fix solution file: update benchmark project path and folder name

fdcastel committed:
Add LargeFetchBenchmark for bulk read throughput across data types

fdcastel committed:
Extract shared BenchmarkConfig to eliminate duplication

fdcastel committed:
Extract BenchmarkBase with shared ConnectionString and env-var override

fdcastel committed:
Update benchmark toolchain from .NET 8 to .NET 10

fdcastel committed:
Add GitHub Markdown exporter and fastest-to-slowest ordering

fdcastel committed:
Replace single-value [Params] with constants

fdcastel committed:
Return values from Fetch benchmarks to prevent dead-code elimination

fdcastel committed:
Add async benchmark variants for Execute and Fetch

fdcastel committed:
Add connection open/close benchmark

fdcastel committed:
Add -Disasm and -Profile flags to run-benchmark.ps1

fdcastel committed:
Fix build errors: remove duplicate braces, add missing using

fdcastel commented:
Following up on the initial refactoring, these additional changes were made on the branch:

**Infrastructure**

- Extracted a shared `BenchmarkConfig : ManualConfig` class to eliminate the duplicated inner `Config` class that existed independently in each benchmark class. Job configuration is now defined in a single place.
- Extracted a `BenchmarkBase` abstract class holding the `ConnectionString`, `CreateDatabase()`, and `GlobalCleanup()` members that were duplicated across benchmark classes.
- The connection string can now be overridden via the `FIREBIRD_BENCHMARK_CS` environment variable (falls back to the existing `localhost` default), enabling use in environments with a non-default Firebird setup.

**Configuration**

- Updated the benchmark toolchain from .NET 8 to .NET 10, matching the project's `TargetFramework`.
- Removed `WithPlatform(Platform.X64)` and `WithJit(Jit.RyuJit)` — both are no-ops on .NET 8+ (x64 + RyuJIT is the only supported configuration) and were just noise.
- Added `MarkdownExporter.GitHub` so each run automatically writes a GitHub-flavored Markdown table to `BenchmarkDotNet.Artifacts/`, ready to paste into issues or PRs.
- Added `DefaultOrderer(SummaryOrderPolicy.FastestToSlowest)` to make the results table easier to scan.

**Benchmark methods**

- Single-value `[Params(100)]` and `[Params(100_000)]` replaced with plain constants — a single-value `[Params]` adds a column to the results table without varying anything.
- `Fetch` methods now return the last read value instead of discarding it, to make dead-code elimination by the JIT impossible.
- Added async variants (`ExecuteAsync`, `FetchAsync`, `OpenCloseAsync`) for all sync benchmarks. Async ADO.NET paths have different allocation profiles (state machines, `Task` wrappers) and are the primary path in web workloads.

**New benchmarks**

- `ConnectionBenchmark` — measures connection open/close throughput (sync and async), which exercises connection pool logic.
- `LargeFetchBenchmark` — added to the committed tree (was previously an untracked file).

**Script**

- `run-benchmark.ps1` now accepts `-Benchmark ConnectionBenchmark` in addition to the existing values, and two new optional switches: `-Disasm` (enables `--disasm`) and `-Profile` (enables ETW profiling on Windows).

**Documentation**

- Added `docs/benchmark.md` describing prerequisites, how to run, connection string configuration, available benchmarks, and where results are written.

fdcastel committed:
Add benchmark documentation

fdcastel commented:
Future improvements: 

1. async benchmarks
2. connection-pool benchmarks
3. configurable connection string
4. Remove Duplicated config (use a shared base)

fdcastel commented:
Added a couple of benchmark improvements prompted by #1272 (the UTF8 `CHAR` fetch allocation regression). The lesson from that investigation: the regression lives only in the fixed-length `CHAR(n) CHARACTER SET UTF8` rune path, but the default `CommandBenchmark` only covered `BIGINT` and `VARCHAR(10) UTF8`, so it never surfaced there — it was visible only via the opt-in `LargeFetchBenchmark`. These changes make the default suite catch that class of regression and make the results easier to read.

## What changed (commit 46db5a1)

- **`CommandBenchmark`**: added `CHAR(100) CHARACTER SET UTF8` to the `DataType` params. This covers the read/truncation path in `Fetch`/`FetchAsync` automatically.
- **`CommandBenchmark.Execute`**: `Execute`/`ExecuteAsync` now send a column-sized string (via a small `SampleValue()` helper) instead of a bare integer, so the write/validate rune-count path is exercised too — previously it was never benchmarked.
- **`docs/benchmark.md`**: corrected the NuGet baseline description (it claimed "latest published" but the project pins a specific released version; it now documents the pin plus a bump-per-release note), added an **Interpreting results** section explaining that for these I/O-bound benchmarks the `Alloc Ratio` (not the noisy `Mean`) is the primary regression signal — using #1272 as a worked example — and listed the new data type.

## Effect — the default benchmark now catches #1272

Running `CommandBenchmark` on this branch, local `Core100` vs the released `NuGet100` baseline:

| DataType | Allocated (baseline → local) | Alloc Ratio |
| --- | --- | --- |
| BIGINT — Fetch | 46.3 KB → 50.9 KB | 1.10 |
| VARCHAR(10) UTF8 — Fetch | 50.2 KB → 54.8 KB | 1.09 |
| CHAR(100) UTF8 — Fetch | 187.7 KB → 1805.1 KB | **9.61** |
| CHAR(100) UTF8 — FetchAsync | 209.6 KB → 1825.1 KB | **8.71** |
| CHAR(100) UTF8 — Execute | 409.5 KB → 759.9 KB | **1.86** |
| CHAR(100) UTF8 — ExecuteAsync | 977.9 KB → 1329.1 KB | **1.36** |

Only the `CHAR UTF8` cases regress; `BIGINT` and `VARCHAR` stay at ~1.0, which also matches the root cause (only `DbDataType.Char` goes through the rune-truncation path). These ratios reflect the current provider code on master — once the fix from #1252 lands they should return to ~1.0.

CI is green across FB30/FB40/FB50.
